File: benchmarks/tsb/bench_interpolate.ts
/**
 * Benchmark: Series.interpolate() — linear interpolation over NaN values.
 * Outputs JSON: {"function": "interpolate", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 50;

const data = Array.from({ length: SIZE }, (_, i) => (i % 5 === 0 ? Number.NaN : i * 1.0));
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  s.interpolate({ method: "linear" });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  s.interpolate({ method: "linear" });
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "interpolate",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_interpolate.py
"""Benchmark: Series.interpolate() — linear interpolation over NaN values."""
import json, time
import pandas as pd
import math

SIZE = 100_000
WARMUP = 5
ITERATIONS = 50

data = [float(i) if i % 5 != 0 else math.nan for i in range(SIZE)]
s = pd.Series(data)

for _ in range(WARMUP):
    s.interpolate(method="linear")

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.interpolate(method="linear")
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function":"interpolate","mean_ms":round(total_ms/ITERATIONS,3),"iterations":ITERATIONS,"total_ms":round(total_ms,3)}))
